dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/deduplication/DeduplicationHelper.java

```diff
@@ -59,7 +60,40 @@
 
     public MergeObject generateMergeObject( TrackedEntityInstance original, TrackedEntityInstance duplicate )
     {
-        return null;
+        if ( duplicate == null )
+        {
+            return null;
+        }
+
+        MergeObject mergeObject = new MergeObject();
+
+        Set<String> existingTrackedEntityAttributes = original.getTrackedEntityAttributeValues().stream()
+            .map( teav -> teav.getAttribute().getUid() ).collect( Collectors.toSet() );
+
+        List<String> trackedEntityAttributes = duplicate.getTrackedEntityAttributeValues().stream()
+            .map( teav -> teav.getAttribute().getUid() )
+            .filter( tea -> !existingTrackedEntityAttributes.contains( tea ) )
+            .collect( Collectors.toList() );
+
+        if ( !trackedEntityAttributes.isEmpty() )
+        {
+            mergeObject.setTrackedEntityAttributes( trackedEntityAttributes );
+        }
+
+        Set<String> existingRelationships = original.getRelationshipItems().stream()
```

Comment:
original.getRelationshipItems() would return a list of relationshipItems that reference original, so all the uids you are getting is the same, namely original's uid.

Instead, if we are getting all relationships that are in duplicate but not original, you would do
original.getRelationshipItems().stream().map( ri -> ri.getRelationship().getUid() ). This way we get the uid of the relationship itself, which we can compare.

If the relationship uid is the same, that implies both original and duplicate are part of the relationship, so we skip it :)